_posts/2017-10-26-totw-94.md

```diff
@@ -0,0 +1,131 @@
+some clues about what they mean. For example, you might not know what to make of
+`absl::string_view s(x, y);` if you'd never seen `string_view` before, but
+`absl::string_view s(my_str.data(), my_str.size());` and `absl::string_view
+s("foo");` are much clearer. The trouble with `bool` parameters is that the
+argument at the callsite is very often a literal `true` or `false`, and that
+gives the reader no contextual cues about the meaning of the parameter, as we
+saw in the `ParseCommandLineFlags` example. This problem is compounded if
+there's more than one `bool` parameter, because now you have the additional
+problem of figuring out which parameter is which.
+
+So how can we fix code like that example? One possibility is to do something
+like this:
+
+```c++ {.bad}
+int main(int argc, char* argv[]) {
+  ParseCommandLineFlags(&argc, &argv, false /* preserve flags */);
+}
+```
+
+The drawbacks of this approach are evident: it's not clear if that comment is
+describing the meaning of the parameter or the effect of the argument. In other
+words, is it saying that we're preserving the flags, or is it saying that it's
+false that we're preserving the flags? Even if the comment manages to make that
+clear, there's still a risk that the comment will get out of sync with the code.
+
+A better approach is to specify the name of the parameter in the comment:
+
+```c++
+int main(int argc, char* argv[]) {
+  ParseCommandLineFlags(&argc, &argv, /*remove_flags=*/false);
+}
+```
+
+This is much clearer, and much less likely to fall out of sync with the code.
+[Clang-tidy will even
+check](https://clang.llvm.org/extra/clang-tidy/checks/misc-argument-comment.html)
+that the comment has the correct parameter name. A less ambiguous but longer
+variant is to use an explanatory variable:
+
+```c++
+int main(int argc, char* argv[]) {
+  const bool remove_flags = false;
+  ParseCommandLineFlags(&argc, &argv, remove_flags);
+}
+```
+
+However, explanatory variable names are not checked by the compiler, and so they
+may be erroneous. This is a particular problem when you have multiple `bool`
+parameters, which might be transposed by the caller.
+
+All these approaches also rely on programmers to consistently remember to add
+these comments or variables, and to do so correctly (although clang-tidy will
+check the correctness of parameter-name comments).
+
+In many cases, the best solution is to avoid using `bool` parameters in the
+first place, and use an `enum` instead. For example, `ParseCommandLineFlags`
+could be declared like this:
+
+```c++
+enum ShouldRemoveFlags { kDontRemoveFlags, kRemoveFlags };
+
+void ParseCommandLineFlags(int* argc, char*** argv, ShouldRemoveFlags remove_flags);
+```
+
+so that the call could look like this:
+
+```c++
+int main(int argc, char* argv[]) {
+  ParseCommandLineFlags(&argc, &argv, kDontRemoveFlags);
+}
+```
+
+You can also use an `enum class`, as described in http://abseil.io/tips/86, but
```

Comment:
Prefer actual links to be relative. I'd also omit the "http:" protocol but leave in "abseil.io" if you want the paradigm for ppl to remember links as "abseil.io/tips/n"

[abseil.io/tips/86](tips/86)